### PR TITLE
[WIP] At cpp ops mapping

### PR DESCRIPTION
**Pull request by atuzhykov**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.

### CODE REVIEW

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2D.java [hunk 1]
```diff
@@ -55,10 +53,17 @@
...
+    protected SDVariable weights;
+    protected SDVariable bias;
+
 
     public DepthwiseConv2D(@NonNull SameDiff sameDiff, @NonNull SDVariable input,
                            @NonNull SDVariable weights, SDVariable bias, @NonNull Conv2DConfig conv2DConfig) {
         this(sameDiff, wrapFilterNull(input, weights, bias), conv2DConfig);
+        this.in = input;
+        this.weights = weights;
+        this.bias = bias;
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2D.java]: We shouldn't need the SDVariable in/weight/bias fields.
We can get them again via arg(int) and args() methods.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2D.java [hunk 2]
```diff
@@ -308,7 +313,7 @@ public String opName() {
 
     @Override
     public List<SDVariable> doDiff(List<SDVariable> f1) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        return Arrays.asList(new DepthwiseConv2DBp(sameDiff, f1, this.config).outputVariables());
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2D.java]: We discussed this.
```
in general (for pretty much every op)
every floating point output of forward pass is a gradient input of backprop
every floating point input of forward pass is a gradient output of backprop
i.e., if X = {x0, x1, ..., xN} is inputs to forward pass
and Y = {y0, y1, ..., yM} is output of forward pass
i.e., Y = op(X)
then backprop is:
dX = opBp(X, dY)
```

You're doing this: `dX = opBp(dY)` not `dX = opBp(X, dY)`

Put another way: It doesn't match the C++ op (i.e., INPUT_VARIABLE(x) part)
https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/convo/depthwiseConv2d.cpp#L152-L155
the C++ op expects 3 or 4 inputs:
input, weights, gradO (i.e., arg(0), arg(1), f1.get(0)), OR
input, weights, bias, gradO (i.e., arg(0), arg(1), arg(2), f1.get(0))

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+ */
+@Slf4j
+@Getter
+@NoArgsConstructor
+public class DepthwiseConv2DBp extends DynamicCustomOp {
+
+    protected Conv2DConfig config;
+
+
+    public DepthwiseConv2DBp(SameDiff sameDiff, List<SDVariable> f1, Conv2DConfig config) {
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java]: Bad signature. It _must_ match the C++ op inputs:
https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/convo/depthwiseConv2d.cpp#L152-L155
```
public DepthwiseConv2DBp(@NonNull SameDiff sameDiff, @NonNull SDVariable input, @NonNull SDVariable weights, SDVariable bias, @NonNull SDVariable gradO, @NonNull Conv2DConfig config) {
```

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        onnxMappings.put("dW", new SizeThresholdIntArrayIntIndexAdpater(1, 2, 0));
+        onnxMappings.put("sH", new SizeThresholdIntArrayIntIndexAdpater(0, 2, 0));
+        onnxMappings.put("sW", new SizeThresholdIntArrayIntIndexAdpater(1, 2, 0));
+        onnxMappings.put("isSameMode", new StringEqualsAdapter("SAME"));
+
+        return ret;
+    }
+
+    @Override
+    public Map<String, Map<String, PropertyMapping>> mappingsForFunction() {
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java]: Cut both the attributeAdaptersForFunction and mappingsForFunction.
These are for tensorflow/onnx import only.
None of the backprop operations should have import support (with like 2 exceptions in the whole codebase).

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        map.put("dataFormat", dataFormat);
+
+
+        return ret;
+    }
+
+
+    @Override
+    public String opName() {
+        return "depthwise_conv2d_bp]";
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java]: Bad op name (unnecessary `]`)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    @Override
+    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
+        int n = args().length;
+        Preconditions.checkState(inputDataTypes != null && inputDataTypes.size() == n, "Expected %s input data types for %s, got %s", n, getClass(), inputDataTypes);
+        return Collections.singletonList(inputDataTypes.get(0));
+    }
+
+    @Override
+    public int getNumOutputs(){return 1;}
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DepthwiseConv2DBp.java]: Incorrect. Does not match c++ op.
https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/nn/convo/depthwiseConv2d.cpp#L157-L159
op has either 2 or 3 outputs, depending on 
we can do something like `return args().length == 4 ? 3 : 2`
as args().length == 4 when bias is provided, or length 3 otherwise (input/weights/bias/gradO vs. input/weights/gradO)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+@NoArgsConstructor
+public class EmbeddingLookup extends DynamicCustomOp {
+
+    @Override
+    public String opName() {
+        return "embedding_lookup";
+    }
+
+
+    public EmbeddingLookup(@NonNull SameDiff sameDiff, @NonNull SDVariable in, @NonNull SDVariable indeces) {
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java]: Typo - `indeces` --> `indices`
(Also below)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        super("embedding_lookup", sameDiff, new SDVariable[]{in, indeces});
+    }
+
+    public EmbeddingLookup(@NonNull INDArray in, @NonNull INDArray indeces, INDArray output) {
+        super("embedding_lookup", new INDArray[]{in, indeces}, new INDArray[]{output});
+    }
+
+    public EmbeddingLookup(@NonNull INDArray in, INDArray output, @NonNull int... indeces) {
+        super("embedding_lookup", new INDArray[]{in, Nd4j.create(indeces)}, new INDArray[]{output});
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java]: Nd4j.create(int...) creates an Float/Double NDArray with _shape_ equal to the provided array, which is not what we want here. i.e., Nd4j.create(1,2,3,4,5) creates a rank 5 array with shape [1,2,3,4,5[

Instead: use Nd4j.createFromArray(int...) which creates a 1d integer array shape [array_length] with _values_ (1,2,3,4,5)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/CReluBp.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public String opName() {
+        return "crelu_bp";
+    }
+
+    @Override
+    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
+        Preconditions
+                .checkArgument(dataTypes != null && dataTypes.size() == 2, "Expected exactly 1 input datatypes, got %s", dataTypes);
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/CReluBp.java]: Condition is good, but message is wrong (`exactly 1` --> `exactly 2`)

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java [hunk 10]
```diff
@@ -1986,4 +1987,95 @@ public void testLogSumExp2(){
...
+        SDVariable expected = sd.math.clipByNorm(in, 1e-2,0,1,2).mul(inputArr.length());
+
+        out.markAsLoss();
+
+
+        OpValidation.validate(new TestCase(sd)
+                .expectedOutput("clipbyavgnorm", expected.eval())
+                .gradientCheck(false)
+        );
+    }
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java:2085]: For all of these tests, let's add a standard deviation op after the output. Something like:
```
SDVariable loss = sd.standardDeviation("loss", out, true);
loss.markAsLoss();
```

The reason: we are more likely to detect bugs using stdev than what's being done here.
When a loss function isn't specified, we by default use sum(last_output) where last_output = out = ClipByAvgNorm in this case.

The reason is simple:
if _out_ has shape [2,2] and loss = sum(out), then dL/dOut[i] = 1 * dSum/dOut[i] = 1
i.e., all the loss values dL/dOut[i] are the same, which won't pick up certain types of failures. Whereas with stdev, the gradients dL/dOut[i] will be all different.

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java [hunk 11]
```diff
@@ -1986,4 +1987,95 @@ public void testLogSumExp2(){
...
+
+
+    @Test
+    public void testDepthwiseConv2D(){
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java]: This test should go into LayerOpValidation not TransformOpValidation)

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java [hunk 12]
```diff
@@ -1986,4 +1987,95 @@ public void testLogSumExp2(){
...
+
+        List<SDVariable> grads = new DepthwiseConv2D(sd,in, weights, null, Conv2DConfig.builder()
+                .kH(kernelHeight)
+                .kW(kernelWidth)
+                .sH(strideHeight)
+                .sW(strideWidth)
+                .dataFormat("NCHW")
+                .build()).doDiff(vars_for_grads);
+        System.out.println(grads.get(0).eval());
+//        java.lang.IllegalStateException: Could not find descriptor for op: depthwise_conv2d_bp] - class: org.nd4j.linalg.api.ops.impl.layers.convolution.DepthwiseConv2DBp
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java]: That comment: due to bad op name `depthwise_conv2d_bp]` not `depthwise_conv2d_bp`
Also just use sd.cnn.depthWiseConv2d method instead of creating manually.

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java [hunk 13]
```diff
@@ -1986,4 +1987,95 @@ public void testLogSumExp2(){
...
+    }
+
+    @Test
+    public void testEmbeddingLookup(){
+
+
+        Nd4j.getRandom().setSeed(12345);
+        SameDiff sd = SameDiff.create();
+        SDVariable in = sd.var("in", Nd4j.rand(1024, 64));
+        SDVariable indeces = sd.var("indeces", Nd4j.create(new int[]{2,20}));
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java]: As per earlier comments:
indeces -> indices
Nd4j.create -> Nd4j.createFromArray

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java [hunk 14]
```diff
@@ -1986,4 +1987,95 @@ public void testLogSumExp2(){
...
+        SDVariable res = new EmbeddingLookup(sd,in, indeces).outputVariable();
+        System.out.println(res.eval());
+
+
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java:2132]: https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup?hl=en

for the 1d input case, expected result it should be equivalent to Nd4j.pullRows method
i.e., "get the rows as specified by the indices, then stack them to get a matrix"

**Review by AlexDBlack — CHANGES_REQUESTED** [review]
See comments.
Also don't forget to add the new ops (excluding the backprop/xBp ops) to the kotlin namespaces.
